plugins/search-backend-node/src/IndexBuilder.ts

```diff
@@ -111,46 +111,43 @@ export class IndexBuilder {
 
     Object.keys(this.collators).forEach(type => {
       scheduler.addToSchedule(async () => {
-        // Collate, Decorate, Index.
-        const decorators: DocumentDecorator[] = (
-          this.decorators['*'] || []
-        ).concat(this.decorators[type] || []);
+        // Instantiate the collator.
+        const collator = await this.collators[type].factory.getCollator();
+        this.logger.info(
+          `Collating documents for ${type} via ${this.collators[type].factory.constructor.name}`,
+        );
 
-        this.logger.debug(
-          `Collating documents for ${type} via ${this.collators[type].collate.constructor.name}`,
+        // Instantiate all relevant decorators.
+        const decorators: Transform[] = await Promise.all(
+          (this.decorators['*'] || [])
+            .concat(this.decorators[type] || [])
+            .map(async factory => {
+              const decorator = await factory.getDecorator();
+              this.logger.info(
+                `Attached decorator via ${factory.constructor.name} to ${type} index pipeline.`,
+              );
+              return decorator;
+            }),
         );
-        let documents: IndexableDocument[];
-
-        try {
-          documents = await this.collators[type].collate.execute();
-        } catch (e) {
-          this.logger.error(
-            `Collating documents for ${type} via ${this.collators[type].collate.constructor.name} failed: ${e}`,
-          );
-          return;
-        }
-
-        for (let i = 0; i < decorators.length; i++) {
-          this.logger.debug(
-            `Decorating ${type} documents via ${decorators[i].constructor.name}`,
-          );
-          try {
-            documents = await decorators[i].execute(documents);
-          } catch (e) {
-            this.logger.error(
-              `Decorating ${type} documents via ${decorators[i].constructor.name} failed: ${e}`,
-            );
-            return;
-          }
-        }
-
-        if (!documents || documents.length === 0) {
-          this.logger.debug(`No documents for type "${type}" to index`);
-          return;
-        }
-
-        // pushing documents to index to a configured search engine.
-        await this.searchEngine.index(type, documents);
+
+        // Instantiate the indexer.
+        const indexer = await this.searchEngine.getIndexer(type);
+
+        // Compose collator/decorators/indexer into a pipeline
```

Comment:
this PR is fantastic and it is going to improve things a lot!

I just want to propose something that could simplify things a bit hopefully make things less breaking change :D

So, `Collator` is now a generator and its implementation looks very neat!

However I'm a bit worried about the knowledge that people need to have for customizing their search engine, since they know need to have the concept of what a stream is.

In this direction I think we could simplify the logic by keeping decorators promises.

Since `Collator` is a generator/readable stream, we could do something like this instead:

```javascript
for await (const document of collator) {
  let result = document;
  for(const decorator of decorators) {
     result = await decorator.execute();
  }
  await indexer.write(result);
}
```

Now we are still using streams, but the process looks more iterative, as it was before.
What do you think?

Reply:
This is an interesting suggestion!  Thanks for bringing up, @vinzscam.

Let me see if I'm understanding your suggestion correctly.  Are you proposing that, instead of a `DocumentDecoratorFactory` that instantiates a transform stream, we keep a `DocumentDecorator` around, but update its interface to be...

```ts
interface DocumentDecorator {
  readonly types?: string[];
  execute(document: IndexableDocument): Promise<IndexableDocument | IndexableDocument[] | undefined>;
}
```

And handle the translation between the Collator/Readable and the Indexer/Writable somewhere in `IndexBuilder` (with code similar to what you posted above)?

Or are you thinking the interface would be more like...

```ts
export interface DocumentDecorator {
  readonly types?: string[];
  execute(documents: AsyncGenerator<IndexableDocument>): AsyncGenerator<IndexableDocument>;
}
```

And then the decorator implementation itself would contain code like what you wrote above?

Reply:
sorry my code didn't make any sense 😅. I've edited.
Yes, the first interface.

But I see now that each decorator could return an array of `IndexableDocument`. This could make things harder with my approach 🤔

Reply:
So, something like this should do the trick:

```javascript
for await (const document of collator) {
  let results = [document];

  for(const decorator of decorators) {
     let decoratorResults = [];
     
     for(const result of results) {
       const result = await decorator.execute(result);
       if(result) {
         if(Array.isArray(result)) {
           decoratorResults.concat(result);
         } else {
           decoratorResults.push(result);
         }
       }
     }
     results = decoratorResults;
  }
  for(const finalResult of results) {
     await indexer.write(finalResult);
  }
}
```

as you can notice, the elements that are kept in memory are due to `Decorator::execute`, which could return a potential array of results. BTW the same happen under the hood if we keep the Decorator as transform stream.
But if they usually return a single element the used memory is still constant.

The good side of this approach is that `DecoratorBase` can go away, since it could contain just an asynchronous method and it could potentially be easier to explain to adopters

Reply:
Thanks for the clarification @vinzscam!  I can see the DX benefits of that approach for decorators, for sure.

I have a couple of reservations:  

**Object lifecycle management**: While overall memory _performance_ is unchanged in either the stream or the plain-old-class approach, the mental model of the _lifecycle_ of the underlying class would become fragmented amongst collators/indexers on one hand, and decorators on the other.  Streams are inherently disposable because they can only ever be used once after instantiation.  An instance of this `DocumentDecorator` interface would live on for as long as the backend was up and running.  That either puts extra memory management burden on the implementor, or we end up with a similar factory pattern as in this PR (in which case, the DX ends up being largely the same, except instead of `MyDecorator implements DocumentDecorator {}`, it's `MyDecorator extends DecoratorBase {}`).

**Inherent decorator complexity**: I wish we had more examples of decorators to see general usage patterns in the wild; but anecdotally, the use-case that inspired the need for decorators at Spotify was appending page view data (via analytics) to TechDocs documents.  This process involves loading data up front, then finding/appending data to each relevant document.  I suspect this is a common pattern, and it implies the need for more than just an async function to define a decorator.  We could of course add lifecycle methods to the `DocumentDecorator` interface too, but then again, we'd be trending toward the `DecoratorBase` base class anyway.

So I'm wondering if, to start, we ship this with the streams / base class exposed.  And then, based on usage/feedback, we can offer alternative, improved DX on top?  It's easy to imagine something like...

```ts
// Either an alternative way to register non-stream decorators
indexBuilder.addDecorator({
  types: ['techdocs'],
  lambda: MyDecorator,
});

// Or a utility decorator implemented as a stream under the hood,
// but which exposed a simpler interface to implementors
indexBuilder.addDecorator({
  types: ['techdocs'],
  factory: LambdaDecoratorFactory.withLambdas([
    async (document) => {/* ... */},
  ]),
});
```

Whaddya think?

Reply:
yeah definitely. This PR is a huge step as it is, so :shipit: :shipit: :shipit:!

My intent was to mention another approach on the table, for avoiding having unexpected corner cases when implementing streams. For example, implementing some back-pressure logic (which isn't in place right now), since DecoratorBase could emit more items compared to how many items are received.

Looking forward to starting testing the new decorators! 😁 🚀